infoschema/builder.go

```diff
@@ -242,6 +245,67 @@ func (b *Builder) applyCreateTables(m *meta.Meta, diff *model.SchemaDiff) ([]int
 	return tblIDs, nil
 }
 
+func (b *Builder) applyExchangePartitionWithTable(m *meta.Meta, diff *model.SchemaDiff) ([]int64, error) {
+	tblIDs, err := b.applyTableUpdate(m, diff)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	for _, opt := range diff.AffectedOpts {
+		affectedDiff := &model.SchemaDiff{
+			Version:     diff.Version,
+			Type:        diff.Type,
+			SchemaID:    opt.SchemaID,
+			TableID:     opt.TableID,
+			OldSchemaID: opt.OldSchemaID,
+			OldTableID:  opt.OldTableID,
+		}
+		affectedIDs, err := b.ApplyDiff(m, affectedDiff)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		tblIDs = append(tblIDs, affectedIDs...)
+
+		// handle partition table and table AutoID
+		err = updateAutoIDForExchangePartition(m, affectedDiff.SchemaID, affectedDiff.TableID, diff.SchemaID, diff.OldTableID)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	return tblIDs, nil
+}
+
+func updateAutoIDForExchangePartition(m *meta.Meta, ptSchemaID, ptID, ntSchemaID, ntID int64) error {
+	// partition table auto IDs.
+	ptAutoIDs, err := m.GetAutoIDAccessors(ptSchemaID, ptID).Get()
+	if err != nil {
+		return err
+	}
+	// non-partition table auto IDs.
+	ntAutoIDs, err := m.GetAutoIDAccessors(ntSchemaID, ntID).Get()
```

Comment:
What would happen in this case?
```
CREATE TABLE pt (a int unsigned auto_increment, b varchar(255)) PARTITION BY RANGE (a) (PARTITION p0 VALUES LESS THAN (1000000), partition p1M values less than (2000000));
create table t (a int unsigned auto_increment, b varchar(255));
insert into pt values (1, "1");
insert into t values (2, "2");
insert into t values (4000000, "4M");
delete from t where a = 4000000;
alter table pt exchange partition p0 with table t;
```

Reply:
I put it to the test，RowID is equal to 4005000.